File: python/Day03.py
from utils.Stopwatch import runWithStopwatch

input = open("../input/day03.txt", "r").readlines()


def part1():
    priority_point = 0
    for line in input:
        stripped_line = line.strip()
        middle_index = int(len(stripped_line) / 2)
        unique_first_compartment = toUniqueList(stripped_line[:middle_index])
        unique_second_compartment = toUniqueList(stripped_line[middle_index:])
        for i in unique_first_compartment:
            for j in unique_second_compartment:
                if i == j:
                    priority_point += toPriority(i)
                    break
    print("Total priority point: " + str(priority_point))


def part2():
    priority_point = 0
    for index in range(int(len(input) / 3)):
        group = input[index * 3:(index + 1) * 3]
        for i in toUniqueList(group[0].strip()):
            for j in toUniqueList(group[1].strip()):
                if i == j:
                    for k in toUniqueList(group[2].strip()):
                        if j == k:
                            priority_point += toPriority(k)
                            break

    print("Total priority point: " + str(priority_point))


def toUniqueList(first_compartment):
    return list(set(first_compartment))


def toPriority(char: chr) -> int:
    return ord(str(char).lower()) - ord("a") + 1 + (26 if str(char).isupper() else 0)


if __name__ == '__main__':
    runWithStopwatch(part1)
    runWithStopwatch(part2)

File: kotlin/src/main/kotlin/adventofcode2022/Day03.kt
package adventofcode2022

import adventofcode2022.util.runWithStopwatch
import java.io.File

private val input = File("./input/day03.txt").readLines()

private fun part1() {
    var priorityPoints = 0
    for (line in input) {
        val (firstComponent, secondComponent) = getCompartments(line)
        for (i in firstComponent.distinct()) {
            for (j in secondComponent.distinct()) {
                if (i == j) {
                    priorityPoints += j.toPriority()
                    break
                }
            }
        }
    }
    println("Total priority: $priorityPoints")
}

private fun part2() {
    var priorityPoints = 0
    for (index in 0 until(input.size / 3)) {
        val group = input.slice((index * 3) until (index + 1) * 3)
        for (i in group[0].toCharArray().distinct()) {
            for (j in group[1].toCharArray().distinct()) {
                if (i == j) {
                    for (k in group[2].toCharArray().distinct()) {
                        if (j == k) {
                            priorityPoints += k.toPriority()
                            break
                        }
                    }
                }
            }
        }
    }
    println("Total priority: $priorityPoints")
}

// NOTE: Extension function give good readability
private fun Char.toPriority(): Int {
    return this.lowercaseChar().code - 'a'.code + 1 + 26 * (if (this.isUpperCase()) 1 else 0)
}

// NOTE: Deconstruction give good readability
private fun getCompartments(line: String): Pair<CharArray, CharArray> {
    val length = line.length
    return Pair(
            line.substring(0, length / 2).toCharArray(),
            line.substring(length / 2, length).toCharArray()
    )
}


fun main() {
    runWithStopwatch { part1() }
    runWithStopwatch { part2() }
}